### PR TITLE
MM-18910 Migrate utils/emoticons to typescript

### PR DESCRIPTION
**Pull request by guigui64**
#### Summary
Migrate 'utils/emoticons.jsx' and associated tests to TypeScript

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12492

#### Comments

Had to force the max-nested-callbacks rule for the test file.
*I don't understand why the error got raised on the tsx file when it was not the case on the jsx file...*

### CODE REVIEW

File: utils/emoticons.test.tsx [hunk 1]
```diff
@@ -1,7 +1,9 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import * as Emoticons from 'utils/emoticons.jsx';
+/* eslint max-nested-callbacks: ["error", 10] */
```
Comment (devinbinnie) [inline utils/emoticons.test.tsx]: Hey @guigui64, I've made a fix for the `max-nested-callbacks` issue, which should now be in master. If you update your branch and do a fresh `npm install` it should pull in the new `eslintrc.json` with the fix. Then you can remove this exception.

File: actions/post_actions.jsx [hunk 2]
```diff
@@ -21,7 +21,7 @@ import {
     RHSStates,
     StoragePrefixes,
 } from 'utils/constants';
-import {matchEmoticons} from 'utils/emoticons.jsx';
+import {matchEmoticons} from 'utils/emoticons.tsx';
```
Comment (enahum) [inline actions/post_actions.jsx]: do we need the extension?

File: components/suggestion/emoticon_provider.jsx [hunk 3]
```diff
@@ -10,7 +10,7 @@ import {getEmojiMap, getRecentEmojis} from 'selectors/emojis';
 
 import store from 'stores/redux_store.jsx';
 
-import * as Emoticons from 'utils/emoticons.jsx';
+import * as Emoticons from 'utils/emoticons.tsx';
```
Comment (enahum) [inline components/suggestion/emoticon_provider.jsx]: same here

File: utils/emoticons.tsx [hunk 4]
```diff
@@ -45,10 +45,18 @@ export function matchEmoticons(text) {
     return emojis;
 }
 
-export function handleEmoticons(text, tokens) {
+export function handleEmoticons(
+    text: string,
+    tokens: Map<string, {value: string; originalText: any}>
```
Comment (enahum) [inline utils/emoticons.tsx]: what's the reason to have `originalText` as any? shouldn't this be a string?

File: utils/post_utils.jsx [hunk 5]
```diff
@@ -24,7 +24,7 @@ import MentionableRenderer from 'utils/markdown/mentionable_renderer';
 import * as Utils from 'utils/utils.jsx';
 import {isMobile} from 'utils/user_agent.jsx';
 
-import * as Emoticons from './emoticons.jsx';
+import * as Emoticons from './emoticons.tsx';
```
Comment (enahum) [inline utils/post_utils.jsx]: me being picky

File: utils/text_formatting.jsx [hunk 6]
```diff
@@ -9,7 +9,7 @@ import {formatWithRenderer} from 'utils/markdown';
 import {getEmojiMap} from 'selectors/emojis';
 import store from 'stores/redux_store.jsx';
 
-import * as Emoticons from './emoticons.jsx';
+import * as Emoticons from './emoticons.tsx';
```
Comment (enahum) [inline utils/text_formatting.jsx]: :)

File: utils/emoticons.test.tsx (hunk 1 above)
Comment (guigui64) [reply utils/emoticons.test.tsx]: Thanks, I'm updating the PR right now ;)

Comment (guigui64) [reply utils/emoticons.test.tsx]: Actually, I still get the same issue after having synced...

`> mattermost-webapp@0.0.1 check /home/guillaume/git/mattermost-webapp
> eslint --ignore-pattern node_modules --ignore-pattern dist --ext .js --ext .jsx --ext tsx --ext ts . --quiet


/home/guillaume/git/mattermost-webapp/utils/emoticons.test.tsx
   31:76  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   32:85  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
   38:54  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   43:70  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   48:67  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   53:60  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   58:54  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   63:70  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   68:70  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   73:76  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   78:76  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   83:87  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   88:87  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
   95:31  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
  100:30  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
  105:28  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
  106:37  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
  111:43  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
  117:30  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
  118:37  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
  123:43  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
  128:27  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks
  134:24  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks

/home/guillaume/git/mattermost-webapp/utils/url.test.tsx
  45:27  error  Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks
  45:66  error  Too many nested callbacks (4). Maximum allowed is 2  max-nested-callbacks

✖ 25 problems (25 errors, 0 warnings)
`

File: components/suggestion/emoticon_provider.jsx (hunk 3 above)
Comment (saturninoabril) [reply components/suggestion/emoticon_provider.jsx]: Please remove the extension for consistency.

Comment (saturninoabril) [reply components/suggestion/emoticon_provider.jsx]: @guigui64 If you don't mind, I'll push commit so we can merge.

**Pull request merged**